scripts/validate_rst_title_capitalization.py

```diff
@@ -0,0 +1,272 @@
+            for key in regex:
+                match = re.search(regex[key], lines[line_no])
+                if match is not None:
+                    if line_no >= 2:
+                        if lines[line_no] == lines[line_no - 2]:
+                            if len(lines[line_no]) == len(lines[line_no - 1]):
+                                yield lines[line_no - 1].translate(table), line_no
+                            break
+                    if len(lines[line_no]) >= len(lines[line_no - 1]):
+                        yield lines[line_no - 1].translate(table), line_no
+
+
+def fill_bad_title_dict(rst_file: str) -> None:
+    """
+    Method that fills up the bad_title_dict with incorrectly capitalized headings
+
+    Parameters
+    ----------
+    rst_file : str
+        Directory address of a .rst file as a string
+
+    """
+
+    if rst_file in bad_title_dict:
+        return
+
+    for title, line_number in find_titles(rst_file):
+        if not is_following_capitalization_convention(title):
+            if rst_file not in bad_title_dict:
+                bad_title_dict[rst_file] = [(title, line_number)]
+            else:
+                bad_title_dict[rst_file].append((title, line_number))
+
+
+def find_rst_files(source_paths: List[str]) -> Generator[str, None, None]:
+    """
+    Given the command line arguments of directory paths, this method
+    yields the strings of the .rst file directories that these paths contain
+
+    Parameters
+    ----------
+    source_paths : str
+        List of directories to validate, provided through command line arguments
+
+    Yields
+    -------
+    directory_address : str
+        Directory address of a .rst files found in command line argument directories
+
+    """
+
+    for directory_address in source_paths:
+        if not os.path.exists(directory_address):
+            raise ValueError(
+                "Please enter a valid path, pointing to a valid file/directory."
+            )
+        elif directory_address.endswith(".rst"):
+            yield directory_address
+        else:
+            for (dirpath, _, filenames) in os.walk(directory_address):
+                for file in filenames:
+                    if file.endswith(".rst"):
+                        yield os.path.join(dirpath, file)
+
+
+def main(source_paths: List[str], output_format: str) -> bool:
+    """
+    The main method to print all headings with incorrect capitalization
+
+    Parameters
+    ----------
+    source_paths : str
+        List of directories to validate, provided through command line arguments
+    output_format : str
+        Output format of the script.
+
+    Returns
+    -------
+    number_of_errors : int
+        True if there are headings that are printed, False if not
+
+    """
+
+    number_of_errors: int = 0
+
+    directory_list = find_rst_files(source_paths)
+
+    for filename in directory_list:
+        fill_bad_title_dict(filename)
```

Comment:
This F feels overcomplicated. Just loop over `find_titles` here, and if `title == correct_title_capitalization(title)` yield a tuple with the file name, and the correct and incorrect titles.

I think you can get rid a lot of the code by doing that, and things will be much clearer.